### ISSUE TITLE
Training machine spec recommendation

### ISSUE THREAD

**Issue by jokandre** [ladaapp/lada#43]
First I would like to say that the project works very well on low resources. Thank you very much!
I am able to do inference on 4GB VRAM Nvidia laptop GPU at around 1.5x video time (1 minute video takes around 1.5 minutes to restore).

I would like to train it myself, and even try to help introduce other models related to video restoration.
Can you share your recommendation on hardware specs for training?
I have tried on a cloud machine with 1 Nvidia T4 GPU (15Gb VRAM), however it seems not enough for the 1st stage of basicvsr++ training.
Given the original settings:
```
train_dataloader = dict(
    num_workers=4,
    batch_size=2,
    persistent_workers=False,
    sampler=dict(type='InfiniteSampler', shuffle=True),
    dataset=dict(
        type='MosaicVideoDataset',
        metadata_root_dir=data_root + "/train/crop_unscaled_meta",
        num_frame=30,
        degrade=True,
        use_hflip=True,
        repeatable_random=False,
        random_mosaic_params=True,
        filter_watermark=False,
        filter_nudenet_nsfw=False,
        filter_video_quality=False,
        lq_size=256),
    collate_fn=dict(type='default_collate'))
```

I have tried reducing `workers`, `batch_size` and `num_frame`. 
Then it worked but the estimated training time is around 1.5 days to 3 days. Is it normal?
Also increasing workers didn't seem to make it training faster.

I also couldn't find what `num_frame` does to the training samples. Does it mean that for each video sample it only takes the first 30 frames?
I am training using around 5 1080p videos probably at 60fps.

Can you provide some details on what time of machine you are using?

**Comment by ladaapp** [comment]
Hi. I've trained it using a RTX3090.

`workers` only affects cpu and RAM consumption. Its used by the dataloader to provide the training samples. It has no effects on GPU VRAM OR GPU compute. You probably don't need to increase it. only if you see that your GPU is not fully utilized then it could be its waiting for the dataloaders and increasing the worker count could solve it.

Your understanding of `num_frame` is correct. Only that its taking `num_frame` not only from the beginning but from a random  starting position of the video. 1 would be a still image, so to capture the dynamics of a video and for the model the reference neighboring frames for restoration it shouldn't be too low.

`batch_size` is the number of videos processed in a single training step. So higher batch size == faster training. maybe need to adjust learning rates though.

`num_frame`  and `batch_size` can be adjusted based on your available GPU VRAM.

> Then it worked but the estimated training time is around 1.5 days to 3 days. Is it normal?
I thinks that sounds reasonable. 

> I am training using around 5 1080p videos probably at 60fps.
You will need thousands to train from scratch. The minimum I trained this model on was 10k clips, maybe less also gives okay-ish results but I'd say start with at least a couple hundred.

**Comment by jokandre** [comment]
Thank you for the reply.

I am also curious about the reason why do you train on 2 stages.

For stage one the configuration says max 100k steps, however the stage 2 resumes from iteration 10k according to the command bellow.

```sh
python scripts/training/train-mosaic-restoration-basicvsrpp.py configs/basicvsrpp/mosaic_restoration_generic_stage2.py --load-from experiments/basicvsrpp/mosaic_restoration_generic_stage1/iter_10000_converted.pth
```
Is stage 1 training designed to be 10k steps or it should continue according to the validation loss?

There are also files for stage 2.5 and 2.6 training, but they are not mentioned on the documentation. Are they supposed to be used or are they just your other experiments?

My current goal is just to train and probably overfit it a bit to see if the blurry images disappear. I've got SSIM of 0.90 but the images are still blurry as if out of focus. Have you noticed it during your testing? 

On the published models, what were the benchmarks regarding SSIM and PSNR.
Just trying to know in order to set some expectations.

**Comment by ladaapp** [comment]
first stage training is necessary so that the generator will not produce garbage when training stage2 when we also train in a GAN setting with a discriminator model.

the 10k vs 100k look like a typo. it should load the last checkpoint from stage1, so in default config 100k

the different stage2 configs is more or less how I trained the current model. based on some experimentation. i wouldn't be surprised if you could reduce it to a single config/setting and train just a bit longer.

cannot give you much advise regarding SSIM, PSNR or other metrics. I've trained, and adjusted the code and configs in multiple steps over the last couple months so my numbers would not be useful, more so if you train it on a tiny dataset.
In my experience SSIM, PSNR are not very reliable. Only in the the beginning when they change much, but for fine-tuning I relied mostly on visual comparison...

@ratbunch trained a model recently, maybe he has something to share though

**Comment by ladaapp** [comment]
looks like main questions have been answered - closing
